### DISCUSSION TITLE
Import miio in Alexa Skill causes "No package metadata was found for importlib_metadata"

### DISCUSSION THREAD

**Discussion by palied-havok** [rytilahti/python-miio#1546] [Q&A]
Hello there,

I am trying to implement an Alexa Skill in Python using the Amazon developer console. 
Requirements.txt  looks like this:

> micloud==0.5
> python-miio==0.5.12

The following line causes the error below: 
> from miio import RoborockVacuum

Error:
> [ERROR] PackageNotFoundError: No package metadata was found for importlib_metadata
> Traceback (most recent call last):
>   File "/var/lang/lib/python3.7/imp.py", line 234, in load_module
>     return load_source(name, filename, file)
>   File "/var/lang/lib/python3.7/imp.py", line 171, in load_source
>     module = _load(spec)
>   File "<frozen importlib._bootstrap>", line 696, in _load
>   File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 728, in exec_module
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
>   File "/var/task/lambda_function.py", line 9, in <module>
>     from miio import RoborockVacuum
>   File "/var/task/miio/__init__.py", line 4, in <module>
>     from importlib_metadata import version  # type: ignore
>   File "/var/task/importlib_metadata/__init__.py", line 623, in <module>
>     __version__ = version(__name__)
>   File "/var/task/importlib_metadata/__init__.py", line 585, in version
>     return distribution(distribution_name).version
>   File "/var/task/importlib_metadata/__init__.py", line 558, in distribution
>     return Distribution.from_name(distribution_name)
>   File "/var/task/importlib_metadata/__init__.py", line 215, in from_name
>     raise PackageNotFoundError(name)

It works fine when I use the same import on any py file on my local machine with python-miio installed. 

Any ideas?

Thanks a lot!

**Comment by palied-havok** [comment]
After further inspection it seems like python-miio is not the issue, instead just importing importlib_metadata in an Alexa-hosted Skill is the issue. 
Most likely this is just my lack of knowledge about importlib_metadata and python in general.

Anyway, if someone has an idea off hand, please let me know. Otherwise feel free to delete this thread. 

Thanks

**Reply by rytilahti** [reply]
Yeah, looks like it's about importlib_metadata not finding the version information in the alexa environment. Does the exception text spell out the package name correctly?

The version information is only exposed via `__version__` variable and also for `kasa --version` which are not necessary for any library functionalities, so we could just fallback to set `unknown` or something as a version string if the query fails.

**Reply by palied-havok** [reply]
I think the full spelled out exception text is just "[ERROR] PackageNotFoundError: No package metadata was found for importlib_metadata".
To me it seems like importlib_metadata itself is missing something and it might not be related to miio at all. But I could be wrong.
In fact the problem already occurs when I just use "import importlib_metadata" in the Alexa python script. miio just does the same implicitly when importing it.
So, I am not sure if the fallback you mentioned for miio would help here, since the origin of the issue according to the exception message and my tests is directly in importlib_metadata.

I have checked the code at https://github.com/python/importlib_metadata/blob/main/importlib_metadata/__init__.py
for "def from_name(cls, name: str)" where the exception comes from. But it does not give me more clues. It just does not find the name "importlib_metadata" in a list/dictionary it should have created before.

I also tried to add importlib-metadata or importlib_metadata to the requirements.txt which is used to pip install packages in the Alexa environment. However, that also does not help.

**Reply by rytilahti** [reply]
It is already installed as you are getting the exception from that library. I'll create an issue to guard that `version()` call inside `kasa/__init__.py` not to crash if the information is not available, if you wish, you could report the "importlib_metadata not finding its own metadata" issue to the issue tracker of that project, they probably have an idea what could be a potential fix (if any) for that :-)

**Reply by palied-havok** [reply]
`It is already installed as you are getting the exception from that library`

Oh right - my bad. It is even in the callstack of course ;) 

> if you wish, you could report the "importlib_metadata not finding its own metadata" issue to the issue tracker of that project, they probably have an idea what could be a potential fix (if any) for that :-)

I will do that. Thanks a lot! Just to make sure - do you mean reporting it to https://github.com/python/importlib_metadata and not kasa?

**Reply by rytilahti** [reply]
Ohhh, indeed, importlib_metadata that is! I thought I was reading an issue on the kasa library, as python-miio does not use that library anymore in the current git :-) So if you can install the git version, it should work directly. The library is undergoing some larger changes and because of that it may take a while until I manage to release a new version on pypi.

**Reply by palied-havok** [reply]
Now I see! I got a bit confused why kasa suddenly came up ;)

I just tested something: If I don't use python-miio in requirements.txt and only use the latest importlib_metadata, it works fine. However, forcing the latest importlib_metadata version in requirements.txt AND python-miio does not work. It causes a conflict, because the python-miio package expects an older version of importlib_metadata. 
Anyway, whatever is causing the issue seems to be fixed in a later importlib_metadata version. 

Yes, in fact using the git version python-miio that does not use importlib_metadata anymore could work. But then again the simple Alexa-hosted coding interface only allows getting packages from pypi by defining them in requirements.txt. 
There is still a different way of manually creating the python environment and package for the Alexa skill - this way I could use the latest git version of python-miio. However, I currently don't have much time to look into that and I think I will just wait for your next pypi push.

Are there any timelines for that?

Btw thanks a lot to you and all other contributors for the great work :) 
Using python-miio on my local machine works like a charm and allows me to control my vacuum cleaner to clean specific rooms (this is actually what the disappointing Mi Home Alexa integration lacks - you can only start (= clean all) or stop (without going to the dock) with that integration - so it is somehow pointless). 
The only thing missing on my side now, is to integrate the python-miio vacuum cleaner control to an Alexa skill.
